falcon/testing/client.py

```diff
@@ -2133,8 +2158,166 @@ async def __aexit__(self, exc_type, exc, tb):
         await self._task_req
 
 
+def _prepare_data_fields(data):
+    """Prepare data  fields for request body.
+
+    Args:
+        data: dict or list of tuples with json data from the request
+
+    Returns: list of 2-tuples (field-name(str), value(bytes)
+
+    """
+    fields = []
+    new_fields = []
+    if not isinstance(data, (list, dict)):
+        raise ValueError('Data must not be a list of tuples or dict.')
+    elif isinstance(data, dict):
+        fields = list(data.items())
+    else:
+        fields = list(data)
+
+    # Append data to the other multipart parts
+    for field, val in fields:
+        if isinstance(val, str) or not hasattr(val, '__iter__'):
+            val = [val]
+        for v in val:
+            if v:
+                # v has to come from json serializable obj, so can't be bytes
+                v = str(v)
+                new_fields.append(
+                    (
+                        field.decode('utf-8') if isinstance(field, bytes) else field,
+                        v.encode('utf-8') if isinstance(v, str) else v,
+                    )
+                )
+            else:
+                new_fields.append(
+                    (field.decode('utf-8') if isinstance(field, bytes) else field, b'')
+                )
+
+    return new_fields
+
+
+def _prepare_files(k, v):
+    """Prepare file attributes for body of request form.
+
+    Args:
+        k: (str), file-name
+        v: fileobj or tuple (filename, data, content_type?, headers?)
+
+    Returns: file_name, file_data, file_content_type, file_header
+
+    """
+    file_content_type = None
+    file_data = None
+    file_name = None
+    if not v:
+        raise ValueError(f'No file provided for {k}')
+    if isinstance(v, (tuple, list)):
+        if len(v) == 2:
+            file_name, file_data = v
+        else:
+            file_name, file_data, file_content_type = v
+        if (
+            len(v) == 3
+            and file_content_type
+            and file_content_type.startswith('multipart/mixed')
+        ):
+            file_data, new_header = _encode_files(json_module.loads(file_data.decode()))
+            file_content_type = 'multipart/mixed; ' + (
+                new_header['Content-Type'].split('; ')[1]
+            )
+    else:
+        # if v is not a tuple or iterable it has to be a filelike obj
+        name = getattr(v, 'name', None)
+        if name and isinstance(name, str) and name[0] != '<' and name[-1] != '>':
+            file_name = os.path.basename(name)
+        else:
+            file_name = k
+        file_data = v
+    if hasattr(file_data, 'read'):
+        file_data = file_data.read()
+    return file_name, file_data, file_content_type
+
+
+def _make_boundary():
+    """
+    Create random boundary to be used in multipar/form-data with files.
+    """
+    boundary = binascii.hexlify(os.urandom(16))
```

Comment:
Since `binascii` is a part of the stdlib, there's no problem with using it, but, OTOH, maybe we could simply use `os.urandom(16).hex()`?

No need to decode anything then either, e.g.,
```python
>>> boundary = os.urandom(16).hex()
>>> boundary
'8bd51c7e4520992fabf833c7be73a72b'
```